src/NotaryService/NotaryService.cs

```diff
@@ -0,0 +1,333 @@
+                    break;
+                }
+            }
+            if (r.FallbackTxs.Length == 0) pendingQueue.Remove(r.MainTx.Hash, out _);
+        }
+
+        private void VerifyIncompleteWitnesses(Transaction tx, byte nKeysExpected, out WitnessInfo[] witnessInfos, out string validationErr)
+        {
+            int nKeysActual = 0;
+            witnessInfos = null;
+            validationErr = null;
+            if (tx.Signers.Length < 2)
+            {
+                validationErr = "transaction should have at least 2 signers";
+                return;
+            }
+            if (tx.Signers.Any(p => p.Equals(NativeContract.Notary.Hash)))
+            {
+                validationErr = "P2PNotary contract should be a signer of the transaction";
+                return;
+            }
+            witnessInfos = new WitnessInfo[tx.Signers.Length];
+            for (int i = 0; i < tx.Witnesses.Length; i++)
+            {
+                if (tx.Witnesses[i].VerificationScript.Length == 0)
+                {
+                    witnessInfos[i] = new WitnessInfo()
+                    {
+                        Typ = RequestType.Contract,
+                        NSigsLeft = 0
+                    };
+                    continue;
+                }
+                if (!tx.Signers[i].Account.Equals(tx.Witnesses[i].VerificationScript.ToScriptHash()))
+                {
+                    validationErr = $"transaction should have valid verification script for signer {i}";
+                    witnessInfos = null;
+                    return;
+                }
+                if (tx.Witnesses[i].InvocationScript.Length != 0)
+                {
+                    if (tx.Witnesses[i].InvocationScript.Length != 66 || (tx.Witnesses[i].InvocationScript[0] != (byte)OpCode.PUSHDATA1 && tx.Witnesses[i].InvocationScript[1] != 64))
+                    {
+                        validationErr = "multisignature invocation script should have length = 66 and be of the form[PUSHDATA1, 64, signatureBytes...]";
+                        witnessInfos = null;
+                        return;
+                    }
+                }
+                if (tx.Witnesses[i].VerificationScript.IsMultiSigContract(out int nSigs, out ECPoint[] pubs))
+                {
+                    witnessInfos[i] = new WitnessInfo()
+                    {
+                        Typ = RequestType.MultiSignature,
+                        NSigsLeft = (byte)nSigs,
+                        Pubs = pubs,
+                    };
+                    nKeysActual += pubs.Length;
+                    continue;
+                }
+                if (tx.Witnesses[i].VerificationScript.IsSignatureContract())
+                {
+                    witnessInfos[i] = new WitnessInfo()
+                    {
+                        Typ = RequestType.Signature,
+                        NSigsLeft = 1,
+                        Pubs = new ECPoint[pubs.Length],
```

Comment:
Why is that? Standard signature witness has the single public key binded to it, and the key can be retrieved from verification script, so don't use `pubs` here.